Extended derived types can't access ancestor members

```
program main
  type ::t1
    integer :: t1mem
  end type t1
  type, extends(t1) :: t2
    integer :: t2mem
  end type t2
  type(t2) :: x2
  print *, x2%t1mem
end program main
```

elicits

```
../../test/component.f90:9:15: error: Component 't1mem' not found in derived type 't2'
    print *, x2%t1mem
                ^^^^^
../../test/component.f90:5:24: Declaration of 't2'
    type, extends(t1) :: t2
                         ^^
```

--- a/lib/semantics/mod-file.cc
+++ b/lib/semantics/mod-file.cc
@@ -272,8 +272,10 @@ std::vector<const Symbol *> CollectSymbols(const Scope &scope) {
   sorted.reserve(scope.size());
   for (const auto &pair : scope) {
     auto *symbol{pair.second};
-    if (symbols.insert(symbol).second) {
-      sorted.push_back(symbol);
+    if (!symbol->test(Symbol::Flag::ParentComp)) {
+      if (symbols.insert(symbol).second) {
+        sorted.push_back(symbol);
+      }
     }
   }
   std::sort(sorted.begin(), sorted.end(), [](const Symbol *x, const Symbol *y) {

--- a/lib/semantics/resolve-names.cc
+++ b/lib/semantics/resolve-names.cc
@@ -214,6 +214,9 @@ public:
   // Emit a message and attached message with two names and locations.
   void Say2(const SourceName &, MessageFixedText &&, const SourceName &,
       MessageFixedText &&);
+  // As above, but first message has a second argument.
+  void Say2(const SourceName &, MessageFixedText &&, const SourceName &,
+      const SourceName &, MessageFixedText &&);
   void Annex(parser::Messages &&);
 
 private:
@@ -326,7 +329,9 @@ public:
     if (it == currScope().end()) {
       const auto pair{currScope().try_emplace(name, attrs, details)};
       CHECK(pair.second);  // name was not found, so must be able to add
-      return *pair.first->second;
+      auto &symbol{*pair.first->second};
+      symbol.add_occurrence(name);
+      return symbol;
     }
     auto &symbol{*it->second};
     symbol.add_occurrence(name);
@@ -575,26 +580,23 @@ private:
   // the interface name, if any.
   const SourceName *interfaceName_{nullptr};
 
-  // Handle a statement that sets an attribute on a list of names.
   bool HandleAttributeStmt(Attr, const std::list<parser::Name> &);
   Symbol &HandleAttributeStmt(Attr, const SourceName &);
-  void DeclareObjectEntity(const parser::Name &, Attrs);
-  void DeclareProcEntity(const parser::Name &, Attrs, const ProcInterface &);
+  void DeclareObjectEntity(const SourceName &, Attrs);
+  void DeclareProcEntity(const SourceName &, Attrs, const ProcInterface &);
   bool ConvertToProcEntity(Symbol &);
-
-  // Set the type of an entity or report an error.
-  void SetType(
-      const SourceName &name, Symbol &symbol, const DeclTypeSpec &type);
+  void SetType(const SourceName &, Symbol &, const DeclTypeSpec &);
   const Symbol *ResolveDerivedType(const SourceName &);
   bool CanBeTypeBoundProc(const Symbol &);
   Symbol *FindExplicitInterface(const SourceName &);
   Symbol &MakeTypeSymbol(const SourceName &, const Details &);
+  bool OkToAddComponent(const SourceName &, bool isParentComp = false);
 
   // Declare an object or procedure entity.
   // T is one of: EntityDetails, ObjectEntityDetails, ProcEntityDetails
   template<typename T>
-  Symbol &DeclareEntity(const parser::Name &name, Attrs attrs) {
-    Symbol &symbol{MakeSymbol(name.source, attrs)};
+  Symbol &DeclareEntity(const SourceName &name, Attrs attrs) {
+    Symbol &symbol{MakeSymbol(name, attrs)};
     if (symbol.has<T>()) {
       // OK
     } else if (symbol.has<UnknownDetails>()) {
@@ -606,23 +608,23 @@ private:
             symbol.has<ProcEntityDetails>())) {
       // OK
     } else if (auto *details{symbol.detailsIf<UseDetails>()}) {
-      Say(name.source,
+      Say(name,
           "'%s' is use-associated from module '%s' and cannot be re-declared"_err_en_US,
-          name.source, details->module().name());
+          name, details->module().name());
     } else if (auto *details{symbol.detailsIf<SubprogramNameDetails>()}) {
       if (details->kind() == SubprogramKind::Module) {
-        Say2(name.source,
+        Say2(name,
             "Declaration of '%s' conflicts with its use as module procedure"_err_en_US,
             symbol.name(), "Module procedure definition"_en_US);
       } else if (details->kind() == SubprogramKind::Internal) {
-        Say2(name.source,
+        Say2(name,
             "Declaration of '%s' conflicts with its use as internal procedure"_err_en_US,
             symbol.name(), "Internal procedure definition"_en_US);
       } else {
         CHECK(!"unexpected kind");
       }
     } else {
-      SayAlreadyDeclared(name.source, symbol);
+      SayAlreadyDeclared(name, symbol);
     }
     return symbol;
   }
@@ -646,9 +648,7 @@ public:
   using SubprogramVisitor::Post;
   using SubprogramVisitor::Pre;
 
-  ResolveNamesVisitor(Scope &rootScope) {
-    PushScope(rootScope);
-  }
+  ResolveNamesVisitor(Scope &rootScope) { PushScope(rootScope); }
 
   // Default action for a parse tree node is to visit children.
   template<typename T> bool Pre(const T &) { return true; }
@@ -701,7 +701,9 @@ private:
   const Symbol *ResolveCoindexedNamedObject(
       const parser::CoindexedNamedObject &);
   const Symbol *ResolveDataRef(const parser::DataRef &);
-  const Symbol *FindComponent(const Symbol &base, const SourceName &component);
+  const Symbol *FindComponent(const Symbol &, const SourceName &);
+  Symbol *FindComponent(const Scope &, const SourceName &);
+  bool CheckAccessibleComponent(const Symbol &);
   void CheckImports();
   void CheckImport(const SourceName &, const SourceName &);
 };
@@ -994,6 +996,11 @@ void MessageHandler::SayAlreadyDeclared(
 void MessageHandler::Say2(const SourceName &name1, MessageFixedText &&msg1,
     const SourceName &name2, MessageFixedText &&msg2) {
   Say(name1, std::move(msg1)).Attach(name2, msg2, name2.ToString().c_str());
+}
+void MessageHandler::Say2(const SourceName &name1, MessageFixedText &&msg1,
+    const SourceName &arg2, const SourceName &name2, MessageFixedText &&msg2) {
+  Say(name1, std::move(msg1), name1, arg2)
+      .Attach(name2, msg2, name2.ToString().c_str());
 }
 void MessageHandler::Annex(parser::Messages &&msgs) {
   messages_.Annex(std::move(msgs));
@@ -1329,13 +1336,14 @@ void ModuleVisitor::AddUse(const SourceName &location,
   if (!useModuleScope_) {
     return;  // error occurred finding module
   }
-  const auto it{useModuleScope_->find(useName)};
+  auto it{useModuleScope_->find(useName)};
   if (it == useModuleScope_->end()) {
     Say(useName, "'%s' not found in module '%s'"_err_en_US, useName,
         useModuleScope_->name());
     return;
   }
-  const Symbol &useSymbol{*it->second};
+  Symbol &useSymbol{*it->second};
+  useSymbol.add_occurrence(useName);
   if (useSymbol.attrs().test(Attr::PRIVATE)) {
     Say(useName, "'%s' is PRIVATE in '%s'"_err_en_US, useName,
         useModuleScope_->name());
@@ -1913,12 +1921,12 @@ bool DeclarationVisitor::CheckUseError(
 
 void DeclarationVisitor::Post(const parser::DimensionStmt::Declaration &x) {
   const auto &name{std::get<parser::Name>(x.t)};
-  DeclareObjectEntity(name, Attrs{});
+  DeclareObjectEntity(name.source, Attrs{});
 }
 
 void DeclarationVisitor::Post(const parser::EntityDecl &x) {
   // TODO: may be under StructureStmt
-  const auto &name{std::get<parser::ObjectName>(x.t)};
+  const auto &name{std::get<parser::ObjectName>(x.t).source};
   // TODO: CoarraySpec, CharLength, Initialization
   Attrs attrs{attrs_ ? *attrs_ : Attrs{}};
   if (!arraySpec().empty()) {
@@ -1926,7 +1934,7 @@ void DeclarationVisitor::Post(const parser::EntityDecl &x) {
   } else {
     Symbol &symbol{DeclareEntity<EntityDetails>(name, attrs)};
     if (auto &type{GetDeclTypeSpec()}) {
-      SetType(name.source, symbol, *type);
+      SetType(name, symbol, *type);
     }
     if (attrs.test(Attr::EXTERNAL)) {
       ConvertToProcEntity(symbol);
@@ -1989,6 +1997,7 @@ bool DeclarationVisitor::Pre(const parser::ValueStmt &x) {
 bool DeclarationVisitor::Pre(const parser::VolatileStmt &x) {
   return HandleAttributeStmt(Attr::VOLATILE, x.v);
 }
+// Handle a statement that sets an attribute on a list of names.
 bool DeclarationVisitor::HandleAttributeStmt(
     Attr attr, const std::list<parser::Name> &names) {
   for (const auto &name : names) {
@@ -2033,11 +2042,11 @@ bool DeclarationVisitor::ConvertToProcEntity(Symbol &symbol) {
 void DeclarationVisitor::Post(const parser::ObjectDecl &x) {
   CHECK(objectDeclAttr_.has_value());
   const auto &name{std::get<parser::ObjectName>(x.t)};
-  DeclareObjectEntity(name, Attrs{*objectDeclAttr_});
+  DeclareObjectEntity(name.source, Attrs{*objectDeclAttr_});
 }
 
 void DeclarationVisitor::DeclareProcEntity(
-    const parser::Name &name, Attrs attrs, const ProcInterface &interface) {
+    const SourceName &name, Attrs attrs, const ProcInterface &interface) {
   Symbol &symbol{DeclareEntity<ProcEntityDetails>(name, attrs)};
   if (auto *details{symbol.detailsIf<ProcEntityDetails>()}) {
     if (interface.type()) {
@@ -2052,11 +2061,11 @@ void DeclarationVisitor::DeclareProcEntity(
 }
 
 void DeclarationVisitor::DeclareObjectEntity(
-    const parser::Name &name, Attrs attrs) {
+    const SourceName &name, Attrs attrs) {
   Symbol &symbol{DeclareEntity<ObjectEntityDetails>(name, attrs)};
   if (auto *details{symbol.detailsIf<ObjectEntityDetails>()}) {
     if (auto &type{GetDeclTypeSpec()}) {
-      SetType(name.source, symbol, *type);
+      SetType(name, symbol, *type);
     }
     if (!arraySpec().empty()) {
       if (!details->shape().empty()) {
@@ -2149,11 +2158,19 @@ bool DeclarationVisitor::Pre(const parser::DerivedTypeStmt &x) {
 void DeclarationVisitor::Post(const parser::DerivedTypeStmt &x) {
   auto &name{std::get<parser::Name>(x.t).source};
   auto &symbol{MakeSymbol(name, GetAttrs(), DerivedTypeDetails{})};
-  auto &details{symbol.get<DerivedTypeDetails>()};
   PushScope(Scope::Kind::DerivedType, &symbol);
-  if (derivedTypeInfo_.extends) {
-    if (auto *extends{ResolveDerivedType(*derivedTypeInfo_.extends)}) {
-      details.set_extends(extends);
+  if (auto *extendsName{derivedTypeInfo_.extends}) {
+    if (auto *extends{ResolveDerivedType(*extendsName)}) {
+      symbol.get<DerivedTypeDetails>().set_extends(extends);
+      // Declare the "parent component"; private if the type is
+      if (OkToAddComponent(*extendsName, true)) {
+        auto &comp{DeclareEntity<ObjectEntityDetails>(*extendsName, Attrs{})};
+        comp.attrs().set(Attr::PRIVATE, extends->attrs().test(Attr::PRIVATE));
+        comp.set(Symbol::Flag::ParentComp);
+        auto &derivedTypeSpec{currScope().MakeDerivedTypeSpec(*extendsName)};
+        derivedTypeSpec.set_scope(currScope());
+        comp.SetType(DeclTypeSpec{DeclTypeSpec::TypeDerived, derivedTypeSpec});
+      }
     }
   }
   EndAttrs();
@@ -2195,13 +2212,15 @@ bool DeclarationVisitor::Pre(const parser::SequenceStmt &x) {
   return false;
 }
 void DeclarationVisitor::Post(const parser::ComponentDecl &x) {
-  const auto &name{std::get<parser::Name>(x.t)};
+  const auto &name{std::get<parser::Name>(x.t).source};
   auto attrs{GetAttrs()};
   if (derivedTypeInfo_.privateComps &&
       !attrs.HasAny({Attr::PUBLIC, Attr::PRIVATE})) {
     attrs.set(Attr::PRIVATE);
   }
-  DeclareObjectEntity(name, attrs);
+  if (OkToAddComponent(name)) {
+    DeclareObjectEntity(name, attrs);
+  }
   ClearArraySpec();
 }
 bool DeclarationVisitor::Pre(const parser::ProcedureDeclarationStmt &) {
@@ -2226,7 +2245,7 @@ void DeclarationVisitor::Post(const parser::ProcInterface &x) {
 }
 
 void DeclarationVisitor::Post(const parser::ProcDecl &x) {
-  const auto &name{std::get<parser::Name>(x.t)};
+  const auto &name{std::get<parser::Name>(x.t).source};
   ProcInterface interface;
   if (interfaceName_) {
     if (auto *symbol{FindExplicitInterface(*interfaceName_)}) {
@@ -2293,6 +2312,7 @@ void DeclarationVisitor::Post(const parser::FinalProcedureStmt &x) {
   }
 }
 
+// Set the type of an entity or report an error.
 void DeclarationVisitor::SetType(
     const SourceName &name, Symbol &symbol, const DeclTypeSpec &type) {
   auto *prevType{symbol.GetType()};
@@ -2385,6 +2405,42 @@ Symbol &DeclarationVisitor::MakeTypeSymbol(
   }
 }
 
+// Return true if it is ok to declare this component in the current scope.
+// Otherwise, emit an error and return false.
+bool DeclarationVisitor::OkToAddComponent(
+    const SourceName &name, bool isParentComp) {
+  const Scope *scope{&currScope()};
+  for (bool inParent{false};; inParent = true) {
+    CHECK(scope->kind() == Scope::Kind::DerivedType);
+    auto it{scope->find(name)};
+    if (it != scope->end()) {
+      Symbol &prev{*it->second};
+      parser::MessageFixedText msg{""_en_US};
+      if (isParentComp) {
+        msg = "Type cannot be extended as it has a component named"
+              " '%s'"_err_en_US;
+      } else if (prev.test(Symbol::Flag::ParentComp)) {
+        msg = "'%s' is a parent type of this type and so cannot be"
+              " a component"_err_en_US;
+      } else if (inParent) {
+        msg = "Component '%s' is already declared in a parent of this"
+              " derived type"_err_en_US;
+      } else {
+        msg = "Component '%s' is already declared in this"
+              " derived type"_err_en_US;
+      }
+      Say2(name, std::move(msg), prev.name(),
+          "Previous declaration of '%s'"_en_US);
+      return false;
+    }
+    auto *extends{scope->symbol()->get<DerivedTypeDetails>().extends()};
+    if (!extends) {
+      return true;
+    }
+    scope = extends->scope();
+  }
+}
+
 // ResolveNamesVisitor implementation
 
 bool ResolveNamesVisitor::Pre(const parser::CommonBlockObject &x) {
@@ -2453,7 +2509,7 @@ bool ResolveNamesVisitor::Pre(const parser::ImportStmt &x) {
 
 bool ResolveNamesVisitor::Pre(const parser::StructureComponent &x) {
   ResolveStructureComponent(x);
-  return true;
+  return false;
 }
 
 const Symbol *ResolveNamesVisitor::ResolveStructureComponent(
@@ -2504,12 +2560,10 @@ const Symbol *ResolveNamesVisitor::ResolveDataRef(const parser::DataRef &x) {
 const Symbol *ResolveNamesVisitor::FindComponent(
     const Symbol &base, const SourceName &component) {
   std::optional<DeclTypeSpec> type;
-  if (auto *details{base.detailsIf<EntityDetails>()}) {
-    type = details->type();
-  } else if (auto *details{base.detailsIf<ObjectEntityDetails>()}) {
+  if (auto *details{base.detailsIf<ObjectEntityDetails>()}) {
     type = details->type();
   } else {
-    Say2(base.occurrences().back(),
+    Say2(base.lastOccurrence(),
         "'%s' is not an object of derived type"_err_en_US, base.name(),
         "Declaration of '%s'"_en_US);
     return nullptr;
@@ -2519,10 +2573,10 @@ const Symbol *ResolveNamesVisitor::FindComponent(
   }
   if (type->category() != DeclTypeSpec::TypeDerived) {
     if (base.test(Symbol::Flag::Implicit)) {
-      Say(base.occurrences().back(),
+      Say(base.lastOccurrence(),
           "'%s' is not an object of derived type; it is implicitly typed"_err_en_US);
     } else {
-      Say2(base.occurrences().back(),
+      Say2(base.lastOccurrence(),
           "'%s' is not an object of derived type"_err_en_US, base.name(),
           "Declaration of '%s'"_en_US);
     }
@@ -2532,19 +2586,52 @@ const Symbol *ResolveNamesVisitor::FindComponent(
   const Scope *scope{derivedTypeSpec.scope()};
   if (!scope) {
     return nullptr;  // previously failed to resolve type
-  }
-  auto it{scope->find(component)};
-  if (it == scope->end()) {
-    auto &typeName{scope->symbol()->name()};
-    Say(component, "Component '%s' not found in derived type '%s'"_err_en_US,
-        component, typeName)
-        .Attach(
-            typeName, "Declaration of '%s'"_en_US, typeName.ToString().data());
+  } else if (auto *result{FindComponent(*scope, component)}) {
+    result->add_occurrence(component);
+    return CheckAccessibleComponent(*result) ? result : nullptr;
+  } else {
+    auto &typeName{scope->name()};
+    Say2(component, "Component '%s' not found in derived type '%s'"_err_en_US,
+        typeName, typeName, "Declaration of '%s'"_en_US);
     return nullptr;
   }
-  auto *symbol{it->second};
-  symbol->add_occurrence(component);
-  return symbol;
+}
+
+// Check that component is accessible from current scope.
+bool ResolveNamesVisitor::CheckAccessibleComponent(const Symbol &component) {
+  if (!component.attrs().test(Attr::PRIVATE)) {
+    return true;
+  }
+  CHECK(component.owner().kind() == Scope::Kind::DerivedType);
+  // component must be in a module/submodule because of PRIVATE:
+  const Scope &moduleScope{component.owner().parent()};
+  CHECK(moduleScope.kind() == Scope::Kind::Module);
+  for (auto *scope{&currScope()}; scope->kind() != Scope::Kind::Global;
+       scope = &scope->parent()) {
+    if (scope == &moduleScope) {
+      return true;
+    }
+  }
+  Say2(component.lastOccurrence(),
+      "PRIVATE component '%s' is only accessible within module '%s'"_err_en_US,
+      moduleScope.name(), component.name(), "Declaration of '%s'"_en_US);
+  return false;
+}
+
+// Look in this type's scope and then its parents for component.
+Symbol *ResolveNamesVisitor::FindComponent(
+    const Scope &type, const SourceName &component) {
+  CHECK(type.kind() == Scope::Kind::DerivedType);
+  auto it{type.find(component)};
+  if (it != type.end()) {
+    return it->second;
+  }
+  auto &details{type.symbol()->get<DerivedTypeDetails>()};
+  if (auto *extends{details.extends()}) {
+    return FindComponent(*extends->scope(), component);
+  } else {
+    return nullptr;
+  }
 }
 
 void ResolveNamesVisitor::Post(const parser::ProcedureDesignator &x) {

--- a/lib/semantics/symbol.h
+++ b/lib/semantics/symbol.h
@@ -272,7 +272,7 @@ public:
   ENUM_CLASS(Flag,
       Function,  // symbol is a function
       Subroutine,  // symbol is a subroutine
-      Implicit,  // symbol is implicity typed
+      Implicit,  // symbol is implicitly typed
       ModFile,  // symbol came from .mod file
       ParentComp  // symbol is the "parent component" of an extended type
   );

--- a/lib/semantics/symbol.h
+++ b/lib/semantics/symbol.h
@@ -269,11 +269,18 @@ std::string DetailsToString(const Details &);
 
 class Symbol {
 public:
-  ENUM_CLASS(Flag, Function, Subroutine, Implicit, ModFile);
+  ENUM_CLASS(Flag,
+      Function,  // symbol is a function
+      Subroutine,  // symbol is a subroutine
+      Implicit,  // symbol is implicity typed
+      ModFile,  // symbol came from .mod file
+      ParentComp  // symbol is the "parent component" of an extended type
+  );
   using Flags = common::EnumSet<Flag, Flag_enumSize>;
 
   const Scope &owner() const { return *owner_; }
   const SourceName &name() const { return occurrences_.front(); }
+  const SourceName &lastOccurrence() const { return occurrences_.back(); }
   Attrs &attrs() { return attrs_; }
   const Attrs &attrs() const { return attrs_; }
   Flags &flags() { return flags_; }

--- a/lib/semantics/unparse-with-symbols.cc
+++ b/lib/semantics/unparse-with-symbols.cc
@@ -40,7 +40,7 @@ public:
   template<typename T> void Post(const parser::Statement<T> &) {
     currStmt_ = nullptr;
   }
-  void Post(const parser::Name &);
+  void Post(const parser::Name &name) { ProcessName(name, isRef()); }
   void Post(const parser::EndFunctionStmt &) { EndScope(); }
   void Post(const parser::EndModuleStmt &) { EndScope(); }
   void Post(const parser::EndMpSubprogramStmt &) { EndScope(); }
@@ -57,8 +57,10 @@ public:
   void Post(const parser::ProcedureDesignator &) { PostReference(); }
   bool Pre(const parser::DerivedTypeSpec &) { return PreReference(); }
   void Post(const parser::DerivedTypeSpec &) { PostReference(); }
-  bool Pre(const parser::UseStmt &) { return PreReference(); }
-  void Post(const parser::UseStmt &) { PostReference(); }
+  bool Pre(const parser::Rename::Names &);
+  bool Pre(const parser::UseStmt &);
+  bool Pre(const parser::DerivedTypeStmt &);
+
   bool Pre(const parser::ImportStmt &) { return PreReference(); }
   void Post(const parser::ImportStmt &) { PostReference(); }
 
@@ -71,6 +73,7 @@ private:
   symbolMap refs_;  // statement location to symbol referenced there
   std::map<const Symbol *, const char *> symbolToStmt_;  // symbol to def
 
+  void ProcessName(const parser::Name &, bool isRef);
   void EndScope();
   bool PreReference();
   void PostReference();
@@ -94,7 +97,7 @@ void SymbolDumpVisitor::PrintSymbols(const parser::CharBlock &location,
     if (done.insert(symbol).second) {
       Indent(out, indent);
       out << '!' << (isDef ? "DEF"s : "REF"s) << ": ";
-      DumpForUnparse(out, symbol->GetUltimate(), isDef);
+      DumpForUnparse(out, *symbol, isDef);
       out << '\n';
     }
   }
@@ -105,14 +108,35 @@ void SymbolDumpVisitor::Indent(std::ostream &out, int indent) const {
   }
 }
 
-void SymbolDumpVisitor::Post(const parser::Name &name) {
+// Special handling for some nodes that both define and reference names.
+bool SymbolDumpVisitor::Pre(const parser::Rename::Names &x) {
+  ProcessName(std::get<0>(x.t), /*isRef*/ false);
+  ProcessName(std::get<1>(x.t), /*isRef*/ true);
+  return false;
+}
+bool SymbolDumpVisitor::Pre(const parser::UseStmt &x) {
+  ProcessName(x.moduleName, /*isRef*/ true);
+  parser::Walk(x.u, *this);
+  return false;
+}
+bool SymbolDumpVisitor::Pre(const parser::DerivedTypeStmt &x) {
+  PreReference();
+  parser::Walk(std::get<0>(x.t), *this);
+  ProcessName(std::get<1>(x.t), /*isRef*/ false);
+  parser::Walk(std::get<2>(x.t), *this);
+  PostReference();
+  return false;
+}
+
+// If there is a symbol set in this name, add it to refs_ or symbolToStmt_.
+void SymbolDumpVisitor::ProcessName(const parser::Name &name, bool isRef) {
   if (const auto *symbol{name.symbol}) {
     CHECK(currStmt_);
     // If this is the first reference to an implicitly defined symbol,
     // record it as a def.
     bool isImplicit{symbol->test(Symbol::Flag::Implicit) &&
         symbolToStmt_.find(symbol) == symbolToStmt_.end()};
-    if (isRef() && !isImplicit) {
+    if (isRef && !isImplicit) {
       refs_.emplace(currStmt_->begin(), symbol);
     } else {
       symbolToStmt_.emplace(symbol, currStmt_->begin());

--- a/test/semantics/CMakeLists.txt
+++ b/test/semantics/CMakeLists.txt
@@ -58,6 +58,7 @@ set(ERROR_TESTS
   resolve31.f90
   resolve32.f90
   resolve33.f90
+  resolve34.f90
 )
 
 # These test files have expected symbols in the source
@@ -67,6 +68,7 @@ set(SYMBOL_TESTS
   symbol03.f90
   symbol04.f90
   symbol05.f90
+  symbol06.f90
 )
 
 # These test files have expected .mod file contents in the source

--- a/test/semantics/resolve34.f90
+++ b/test/semantics/resolve34.f90
@@ -1,0 +1,107 @@
+! Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+! Extended derived types
+
+module m1
+  type :: t1
+    integer :: x
+    !ERROR: Component 'x' is already declared in this derived type
+    real :: x
+  end type
+end
+
+module m2
+  type :: t1
+    integer :: i
+  end type
+  type, extends(t1) :: t2
+    !ERROR: Component 'i' is already declared in a parent of this derived type
+    integer :: i
+  end type
+end
+
+module m3
+  type :: t1
+  end type
+  type, extends(t1) :: t2
+    integer :: i
+    !ERROR: 't1' is a parent type of this type and so cannot be a component
+    real :: t1
+  end type
+  type, extends(t2) :: t3
+    !ERROR: 't1' is a parent type of this type and so cannot be a component
+    real :: t1
+  end type
+end
+
+module m4
+  type :: t1
+    integer :: t1
+  end type
+  !ERROR: Type cannot be extended as it has a component named 't1'
+  type, extends(t1) :: t2
+  end type
+end
+
+module m5
+  type :: t1
+    integer :: t2
+  end type
+  type, extends(t1) :: t2
+  end type
+  !ERROR: Type cannot be extended as it has a component named 't2'
+  type, extends(t2) :: t3
+  end type
+end
+
+module m6
+  ! t1 can be extended if it is known as anything but t3
+  type :: t1
+    integer :: t3
+  end type
+  type, extends(t1) :: t2
+  end type
+end
+subroutine s6
+  use :: m6, only: t3 => t1
+  !ERROR: Type cannot be extended as it has a component named 't3'
+  type, extends(t3) :: t4
+  end type
+end
+subroutine r6
+  use :: m6, only: t5 => t1
+  type, extends(t5) :: t6
+  end type
+end
+
+module m7
+  type, private :: t1
+    integer :: i1
+  end type
+  type, extends(t1) :: t2
+    integer :: i2
+    integer, private :: i3
+  end type
+end
+subroutine s7
+  use m7
+  type(t2) :: x
+  integer :: j
+  j = x%i2
+  !ERROR: PRIVATE component 'i3' is only accessible within module 'm7'
+  j = x%i3
+  !ERROR: PRIVATE component 't1' is only accessible within module 'm7'
+  j = x%t1%i1
+end

--- a/test/semantics/symbol06.f90
+++ b/test/semantics/symbol06.f90
@@ -1,0 +1,103 @@
+! Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+!DEF: /main MainProgram
+program main
+ !DEF: /main/t1 DerivedType
+ type :: t1
+  !DEF: /main/t1/a1 ObjectEntity INTEGER(4)
+  integer :: a1
+ end type
+ !DEF: /main/t2 DerivedType
+ !REF: /main/t1
+ type, extends(t1) :: t2
+  !DEF: /main/t2/a2 ObjectEntity INTEGER(4)
+  integer :: a2
+ end type
+ !DEF: /main/t3 DerivedType
+ !REF: /main/t2
+ type, extends(t2) :: t3
+  !DEF: /main/t3/a3 ObjectEntity INTEGER(4)
+  integer :: a3
+ end type
+ !DEF: /main/x3 ObjectEntity TYPE(t3)
+ !REF: /main/t3
+ type(t3) :: x3
+ !DEF: /main/i ObjectEntity INTEGER(4)
+ integer i
+ !REF: /main/i
+ !REF: /main/x3
+ !REF: /main/t2/a2
+ i = x3%a2
+ !REF: /main/i
+ !REF: /main/x3
+ !REF: /main/t1/a1
+ i = x3%a1
+ !REF: /main/i
+ !REF: /main/x3
+ !REF: /main/t3/t2
+ !REF: /main/t2/a2
+ i = x3%t2%a2
+ !REF: /main/i
+ !REF: /main/x3
+ !REF: /main/t3/t2
+ !REF: /main/t1/a1
+ i = x3%t2%a1
+ !REF: /main/i
+ !REF: /main/x3
+ !REF: /main/t2/t1
+ !REF: /main/t1/a1
+ i = x3%t1%a1
+ !REF: /main/i
+ !REF: /main/x3
+ !REF: /main/t3/t2
+ !REF: /main/t2/t1
+ !REF: /main/t1/a1
+ i = x3%t2%t1%a1
+end program
+
+!DEF: /m1 Module
+module m1
+ !DEF: /m1/t1 PUBLIC DerivedType
+ type :: t1
+  !DEF: /m1/t1/t1 ObjectEntity INTEGER(4)
+  integer :: t1
+ end type
+end module
+
+!DEF: /s1 Subprogram
+subroutine s1
+ !DEF: /s1/t2 Use
+ !REF: /m1
+ !REF: /m1/t1
+ use :: m1, only: t2 => t1
+ !DEF: /s1/t3 DerivedType
+ !REF: /s1/t2
+ type, extends(t2) :: t3
+ end type
+ !DEF: /s1/x ObjectEntity TYPE(t3)
+ !REF: /s1/t3
+ type(t3) :: x
+ !DEF: /s1/i ObjectEntity INTEGER(4)
+ integer i
+ !REF: /s1/i
+ !REF: /s1/x
+ !REF: /m1/t1/t1
+ i = x%t1
+ !REF: /s1/i
+ !REF: /s1/x
+ !REF: /s1/t3/t2
+ !REF: /m1/t1/t1
+ i = x%t2%t1
+end subroutine

